### PR TITLE
Query: Lift subqueries projecting single scalar values

### LINKED ISSUE #15705: Convert Enumerable methods to queryable methods when appropriate
Especially when Nav Rewrite converts collection navigation to Queryable.

### PR DESCRIPTION
**Pull request by smitpatel**
Resolves #15705

### CODE REVIEW

File: src/EFCore/Query/Pipeline/EnumerableToQueryableReMappingExpressionVisitor.cs [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    {
+        protected override Expression VisitMethodCall(MethodCallExpression methodCallExpression)
+        {
+            if (methodCallExpression.Method.DeclaringType == typeof(Enumerable))
+            {
+                var arguments = VisitAndConvert(methodCallExpression.Arguments, nameof(VisitMethodCall));
+                var enumerableMethod = methodCallExpression.Method;
+                var enumerableParameters = enumerableMethod.GetParameters();
+                Type[] genericArguments = null;
+                if (enumerableMethod.Name == nameof(Enumerable.Min)
```
Comment (maumar) [inline src/EFCore/Query/Pipeline/EnumerableToQueryableReMappingExpressionVisitor.cs]: maybe add comment why we do this, it seems weird to special case Min/Max unless one knows the underlying problem

File: src/EFCore.Relational/Query/Pipeline/SqlExpressions/SubSelectExpression.cs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+                return hashCode;
+            }
+        }
+
+        #endregion
+
+        public override void Print(ExpressionPrinter expressionPrinter)
+        {
+            expressionPrinter.StringBuilder.Append("(");
+            expressionPrinter.Visit(Subquery);
```
Comment (maumar) [inline src/EFCore.Relational/Query/Pipeline/SqlExpressions/SubSelectExpression.cs]: add indent

Comment (smitpatel) [reply src/EFCore.Relational/Query/Pipeline/SqlExpressions/SubSelectExpression.cs]: What do you mean?

File: src/EFCore.Relational/Query/Pipeline/PipeLine/RelationalSqlTranslatingExpressionVisitor.cs [hunk 3]
```diff
@@ -73,27 +79,73 @@ protected override Expression VisitMember(MemberExpression memberExpression)
...
+                        || methodCallExpression.Method.Name == nameof(Queryable.Contains))
+                    {
+                        if (subquery.Tables.Count == 0
+                            && subquery.Projection.Count == 1)
+                        {
+                            return subquery.Projection[0].Expression;
+                        }
+                        else
+                        {
+                            throw new InvalidOperationException();
```
Comment (maumar) [inline src/EFCore.Relational/Query/Pipeline/PipeLine/RelationalSqlTranslatingExpressionVisitor.cs]: combine the throw with the one below

File: src/EFCore.Relational/Query/Pipeline/PipeLine/QuerySqlGenerator.cs [hunk 4]
```diff
@@ -635,5 +635,14 @@ protected override Expression VisitLeftJoin(LeftJoinExpression leftJoinExpressio
 
             return leftJoinExpression;
         }
+
+        protected override Expression VisitSubSelect(SubSelectExpression subSelectExpression)
+        {
+            _relationalCommandBuilder.Append("(");
+            Visit(subSelectExpression.Subquery);
```
Comment (maumar) [inline src/EFCore.Relational/Query/Pipeline/PipeLine/QuerySqlGenerator.cs]: might want to add indent here also

File: src/EFCore.Relational/Query/Pipeline/PipeLine/RelationalSqlTranslatingExpressionVisitor.cs (hunk 3 above)
Comment (smitpatel) [reply src/EFCore.Relational/Query/Pipeline/PipeLine/RelationalSqlTranslatingExpressionVisitor.cs]: Don't see an easy way.

File: src/EFCore/Query/Pipeline/EnumerableToQueryableReMappingExpressionVisitor.cs (hunk 1 above)
Comment (smitpatel) [reply src/EFCore/Query/Pipeline/EnumerableToQueryableReMappingExpressionVisitor.cs]: I believe it is not too difficult to figure out why we have to special case here. (It's not just extra overloads but no of generic parameters everything also change). While it is not obvious but it has enough pointers. The connected issue can always tell if we forget. Comment would add minimal value.

**Pull request merged**